Add a few python3 fix, mostly found by trial and errors

mscherer opened the pull request:
##### ISSUE TYPE
- Bugfix Pull Request
- Docs Pull Request
##### SUMMARY

So far, with this, I am able to run command remotely on a
python3 only system.

```
 $ ansible  -u root -i 'fedora23,' all -a id -e 'ansible_python_interpreter=/usr/bin/python3'  
fedora23 | SUCCESS | rc=0 >>
uid=0(root) gid=0(root) groupes=0(root) contexte=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
```

However, as that's mostly trial and errors, I am not sure if I am doing the right fixes.

mscherer committed:
Add a few python3 fix, mostly found by trial and errors

So far, with this, I am able to run command remotely on a
python3 only system.

mscherer committed:
Add more fix for python3, so now, setup can run without crash

bcoca commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1599,7 +1601,7 @@ def _log_invocation(self):
                     param_val = str(param_val)
                 elif isinstance(param_val, unicode):
-                    param_val = param_val.encode('utf-8')
+                    param_val = str(param_val.encode('utf-8'))
```
don't you want to_str?

mscherer replied:
Do not seems to be present. But yeah, because this is likely doing something wrong (as I found out on others part of the code)

mscherer replied:
In fact, the problem is that on py3, this creates a bytes object, and so I need to convert it to string again. So i guess the right fix is to not encode in the first place for python3 or something like this.

bcoca commented on lib/ansible/module_utils/basic.py:
```diff
@@ -422,7 +422,7 @@ def remove_values(value, no_log_strings):
         else:
             value = bytes_value
-    elif isinstance(value, Sequence):
+    elif isinstance(value, (Sequence, type({}.keys()))):
```
is this just trying to match the iterable type returned by dict.keys()?

mscherer replied:
Yes, I was pondering on finding a better way, but didn't look further.

sivel replied:
In py3.3 this is accessible from `collections.abc.KeysView`.  I would move this into the `try/except` logic farther up, where `Sequence` is built for older python.  Maybe create `SEQUENCE_TYPES` or such.

mscherer replied:
Seems sensible, let's do that

mscherer commented:
So it fail for now on "test_utf8_output (units.module_utils.basic.test_run_command.TestAnsibleModuleRunCommand) ... ERROR"

mscherer commented on lib/ansible/module_utils/basic.py:
```diff
@@ -2095,11 +2097,15 @@ def run_command(self, args, check_rc=False, close_fds=True, executable=None, dat
                 if cmd.stdout in rfd:
                     dat = os.read(cmd.stdout.fileno(), 9000)
+                    if isinstance(dat, bytes):
+                        dat = dat.decode('utf-8', 'replace')
```
So this part is likely responsible of the test failure.

mscherer committed:
Fix the tests

Since run_command is supposed to return a bytes, we have
to encode, but only on python2, since that's ok on python 3

mscherer committed:
Remove the hack to detect dict_keys, and use KeysView

suggested by @sivel

mscherer commented:
So for now, I do push on the same branch so people can see the evolution, but I will squash the commits for merge (or split them in smaller chunk, if that's better)

mscherer committed:
One more fix found after testing more complex role

abadger commented:
Just took a look at this to see what might have been fixed through other commits.  Looks like facts.py is now fixed.  I think all but these three functions are now fixed in basic.py:
- _check_arguments
- _log_invocation
- run_command

mscherer commented:
Yeah, that's a old PR, not sure if anything is needed anymore, let's close it. If there is problems, they will surface again sooner or later.
